sci-libs/pcl/pcl-9999.ebuild

```diff
@@ -69,37 +64,77 @@ REQUIRED_USE="
 "
 
 PATCHES=(
-	"${FILESDIR}"/${PN}-1.12.1-allow-configuration-of-install-dirs.patch
-	"${FILESDIR}"/${PN}-1.12.1-fix-hardcoded-relative-directory-of-the-installed-cmake-files.patch
+	"${FILESDIR}/${PN}-1.12.1-allow-configuration-of-install-dirs.patch"
+	"${FILESDIR}/${PN}-1.12.1-fix-hardcoded-relative-directory-of-the-installed-cmake-files.patch"
 )
 
-src_prepare() {
-	if use cuda; then
-		cuda_src_prepare
-		cuda_add_sandbox -w
-	fi
+cuda_set_CUDAHOSTCXX() {
+	local compiler
+	tc-is-gcc && compiler="gcc"
+	tc-is-clang && compiler="clang"
+	[[ -z "$compiler" ]] && die "no compiler specified"
+
+	local package="sys-devel/${compiler}"
+	local version="${package}"
+	local CUDAHOSTCXX_test
+	while
+		CUDAHOSTCXX="${CUDAHOSTCXX_test}"
+		version=$(best_version "${version}")
+		if [[ -z "${version}" ]]; then
+			if [[ -z "${CUDAHOSTCXX}" ]]; then
+				die "could not find supported version of ${package}"
+			fi
+			break
+		fi
+		CUDAHOSTCXX_test="$(
+			dirname "$(
+				realpath "$(
+					which "${compiler}-$(echo "${version}" | grep -oP "(?<=${package}-)[0-9]*")"
+				)"
+			)"
+		)"
+		version="<${version}"
+	do ! echo "int main(){}" | nvcc "-ccbin ${CUDAHOSTCXX_test}" - -x cu &>/dev/null; done
 
+	export CUDAHOSTCXX
+}
+
+cuda_get_host_arch() {
+	[[ -z "${CUDAARCHS}" ]] && einfo "trying to determine host CUDAARCHS"
+	: "${CUDAARCHS:=$(__nvcc_device_query)}"
+	einfo "building for CUDAARCHS = ${CUDAARCHS}"
+
+	export CUDAARCHS
+}
+
+src_prepare() {
 	cmake_src_prepare
 }
 
 src_configure() {
+	if use cuda; then
+		cuda_add_sandbox -w
+		cuda_set_CUDAHOSTCXX
+		cuda_get_host_arch
+	fi
+
 	local mycmakeargs=(
 		"-DDOC_INSTALL_DIR=share/doc/${PF}"
 		"-DLIB_INSTALL_DIR=$(get_libdir)"
 		"-DPCLCONFIG_INSTALL_DIR=share/cmake/${PN}-$(ver_cut 1-2)"
-		"-DWITH_CUDA=$(usex cuda TRUE FALSE)"
-		"-DWITH_LIBUSB=$(usex usb TRUE FALSE)"
-		"-DWITH_OPENGL=$(usex opengl TRUE FALSE)"
-		"-DWITH_PNG=$(usex png TRUE FALSE)"
-		"-DWITH_QHULL=$(usex qhull TRUE FALSE)"
-		"-DWITH_QT=$(usex qt5 TRUE FALSE)"
-		"-DWITH_VTK=$(usex vtk TRUE FALSE)"
-		"-DWITH_PCAP=$(usex pcap TRUE FALSE)"
-		"-DWITH_OPENNI=$(usex openni TRUE FALSE)"
-		"-DWITH_OPENNI2=$(usex openni2 TRUE FALSE)"
-		"-DPCL_ENABLE_SSE=$(usex cpu_flags_x86_sse TRUE FALSE)"
-		"-DWITH_DOCS=$(usex doc TRUE FALSE)"
-		"-DWITH_TUTORIALS=$(usex tutorials TRUE FALSE)"
```

Comment:
Actually, we need to be this weirdly specific in `usex`, the build system does literal string comparison here and there, so the default `yes` and `no` won't work.